### ISSUE TITLE
Nested VecElement structures generate bulky code

### ISSUE THREAD

**Issue by dbeach24** [JuliaLang/julia#18857]
Julia 0.5.0 generates inefficient LLVM code when tuples of `VecElement` are composed within other tuples or immutable types.  This is problematic, for example, when trying to leverage SIMD within a dual numbers implementation.

Suppose we define a simple dual number structure as follows:

```jl
immutable Dual{T <: Real} <: Real
    val::T
    partials::NTuple{3,T}
end
```

and a similar `SIMDDual` type as follows:

```jl
immutable SIMDDual{T <: Real} <: Real
    val::T
    partials::NTuple{3,VecElement{T}}
end
```

For each of these, `+` is defined by adding the corresponding `val` and `partials` elements.  (See attached code.)

Using the above, let's define some dual numbers as follows:

```jl
const r = Dual{Float64}(1., (1., 2., 3.))
const s = SIMDDual{Float64}(1.,
    (VecElement(1.), VecElement(2.), VecElement(3.)))

# compositions of dual numbers
# these are the sorts of structures you would use to compute 2nd derivatives
const a = Dual{Dual{Float64}}(r, (r, r, r))
const b = SIMDDual{SIMDDual{Float64}}(s, (s, s, s))
const c = Dual{SIMDDual{Float64}}(s, (s, s, s))
```

When evaluating these codes:
- `@code_llvm a+a` => 117 lines
- `@code_llvm b+b` => 151 lines
- `@code_llvm c+c` => 151 lines

See attached script and output files:
- [simdcheck.jl.txt](https://github.com/JuliaLang/julia/files/518693/simdcheck.jl.txt)
- [a.txt](https://github.com/JuliaLang/julia/files/518694/a.txt)
- [b.txt](https://github.com/JuliaLang/julia/files/518695/b.txt)
- [c.txt](https://github.com/JuliaLang/julia/files/518696/c.txt)

The point here is to show that merely adding the `VecElement` wrapper increases the number of instructions. 

I know I can't expect to get SIMD instructions automatically from the compiler here; *that's not the issue*.  The problem is that merely using the `VecElement` wrapper causes code to be generated that is far more bulky.

Why is this happening?  Can it be fixed?

See also: JuliaDiff/ForwardDiff.jl#98

EDIT:  Updated to fix a bug in the example code and for corrected line counts.

**Comment by yuyichao** [comment]
FWIW all of the three raises method error for me and you should at least fix that before checking performance.

**Comment by yuyichao** [comment]
Also, I don't think this is how `VecElement` is supposed to be used. They won't make things magically better if you are using them as more complicated tuples. There's not support for doing arithmetics on them in Base at all and a better API to actually use them is being experimented at https://github.com/eschnett/SIMD.jl before it's finalized.

**Comment by dbeach24** [comment]
@yuyichao  I spent hours testing, distilling, and writing up this issue.  I really don't appreciate the slap-dash "won't fix" label (before I have even replied!) or being accused of having "magical" expectations.

From the issue description above (emphasis added):

> I know I can't expect to get SIMD instructions automatically from the compiler here; that's not the issue. **The problem is that merely using the `VecElement` wrapper causes code to be generated that is far more bulky.**

I downloaded and re-ran the script I attached.  It runs without error.  If you're getting a method error when running this, then **please share what it is**.

I feel your comment here is inaccurate:

> Also, I don't think this is how VecElement is supposed to be used. They won't make things magically better if you are using them as more complicated tuples.

Look at the example of `c + c`.  In this case, the `VecElement` is only applied at the innermost level, and still the code generation is virtually identical to the `b + b` case (nested `VecElement`s at both levels).  Both of these are far worse than the `a + a` case (without `VecElement`).

Again, this issue isn't about getting proper SIMD code generation from Julia/LLVM; it's about the cost of using `VecElement`.

**Comment by yuyichao** [comment]
> I know I can't expect to get SIMD instructions automatically from the compiler here; that's not the issue. The problem is that merely using the VecElement wrapper causes code to be generated that is far more bulky.

All what I'm saying is that this is simply because this is not how VecElement is supposed to be used and when used this way the effect you see is totally expected. (See below)

> I downloaded and re-ran the script I attached. It runs without error. If you're getting a method error when running this, then please share what it is.

The script doesn't have any error. However, if you actually call any of the methods or look at the results of any of the reflection function (i.e. `code_warntype` or `code_llvm`) instead of just blindly count the number of lines in `code_llvm` you can clearly see that there's error in the code.

> Look at the example of c + c. In this case, the VecElement is only applied at the innermost level, and still the code generation is virtually identical to the b + b case (nested VecElements at both levels). Both of these are far worse than the a + a case (without VecElement).
>
> Again, this issue isn't about getting proper SIMD code generation from Julia/LLVM; it's about the cost of using VecElement.

And again, this is simply because you are not using it in the proper way. `Tuple` of `VecElement` is mapped to llvm vectors which is useful to interact with code that consumes it directly (i.e. `ccall` and `llvmcall`). It is not a better tuple and it does not provide better indexing efficiency. In fact, as what you've seen, indexing into it is worse since that's not how LLVM vectors are supposed to be used.

In short, you should only use VecElement if you want to manually write SIMD code. The way to do it is to write `llvmcall`, `ccall` (of vector math library) or use SIMD.jl. It's possible that we'll add some of the code from SIMD into base (or the standard distribution of packages) once it's more stable. Indexing/using it like a tuple **will** give you worse code/performance.

**Comment by yuyichao** [comment]
> @yuyichao I spent hours testing, distilling, and writing up this issue. I really don't appreciate the slap-dash "won't fix" label (before I have even replied!) or being accused of having "magical" expectations.

Well, sorry about that. If anything, that's because I really dislike people that use low level reflection tools then what they can understand and simply count lines to predict performance. This has been the source of many confusions on the mailing list and in most cases the problem can been seen much more easily with a higher level interface (most common using code_llvm/code_native instead of code_warntype or using code_native instead of code_llvm).

Also the `won't fix` label is not about the error in the code (that's what the `needs more info` for, where the "more info" is a version of code that doesn't throw an method error with `a + a`, `b + b` or `c + c`). The won't fix label is for this not being a valid bug report (i.e. the behavior is expected as a result from incorrect usage and there's nothing to fix). This is more of a usage question than a bug report and should probably go to the mailing list instead. (Or if you can try SIMD.jl and if it doesn't fit your need, an issue with your not-yet-supported usecase there is even better.)

**Comment by dbeach24** [comment]
@yuyichao  Thanks for the recent two remarks.  They're a huge improvement because they're *actionable*.

> Well, sorry about that. If anything, that's because I really dislike people that use low level reflection tools then what they can understand and simply count lines to predict performance. 

Not quite.  I actually had side-by-side diff'd the `@code_llvm` output.  The line counts were just a way to summarize it in the description.  Also, this "example" was based on a much larger body of code where I had discovered this issue via timing tests.  

I must apologize for the broken code.  I clearly made an error when attempting to distill this from a much larger body of code, and messed up somewhere along the way.  I've update the description and files with what should be a fix.  (Apparently the culprit is that `@inbounds` returns `nothing` when applied to an expression...  I had hastily added this and saw that my script still "ran.")

> In short, you should only use VecElement if you want to manually write SIMD code.

Yes, I actually know that.  For a full background, I was trying to make a study of how to introduce SIMD optimizations to `ForwardDiff.jl`.  The intended steps were:

1. Introduce `VecElement` into the `ForwardDiff.Partials{N,T}` type, and get all the tests running.
2. Add the `llvmcall` generation for the case where the partials.

The problem is that, while (1) worked, it introduced huge performance regressions in any tests that involved nested dual numbers.  These performance regressions were so large as to suggest that any performance improvement from introducing `llvmcall` would be ultimately overwhelmed.

**Comment by dbeach24** [comment]
I'm still somewhat confused about the proper use of `VecElement`, and I really want to understand this, so if you can please bear with me...

Consider only the following type:

```jl
immutable Dual{T <: Real} <: Real
    value::T
    partials::NTuple{3, VecElement{T}}
end
```

(Assume here that special `llvmcall` code is added for cases of adding Dual numbers where T is a `bitstype`.)

Where does the use of `VecElement` start to go sideways for me?  (i.e. which of the following types would be considered a misuse/abuse of `VecElement`?)

1. `Dual{Float64}`
2. `Complex{Dual{Float64}}`
3. `Dual{Complex{Float64}}`
4. `NTuple{3, Dual{Float64}}`
5. `Dual{Dual{Float64}}`

**Comment by yuyichao** [comment]
From what I can tell, the main difference in your code comes from using `VecElement` tuple that maps to llvm vectors but do not use vector operations on them and not actually from using a `VecElement` tuple on a type that doesn't map to llvm vectors. In another word, using a `const d = SIMDDual{Dual{Float64}}(r, (r, r, r))` on your fixed code generates very similar code and also has similar performance as `a`. This is why I think your issue is to use LLVM vectors properly.

P.S. `Dual{Complex{Float64}}` probably won't work for unrelated reasons........

**Comment by yuyichao** [comment]
In another word, you see performance issue in the nested case mainly because you are generating more code that trashes llvm optimizations, not because the top level `VecElement`, which doesn't affect the LLVM type, is doing anything bad.

**Comment by dbeach24** [comment]
Ok, so the "outer" tuple trashes the llvm optimizations, but only when the "inner" tuple is `VecElement{T}` ?  (Because of course, if we eliminates the `VecElement`, then everything works swimmingly.)

**Comment by yuyichao** [comment]
More like the inner tuple of `VecElement` trashes llvm when it's not used properly (i.e. not doing vector operations on them). However, with the code being simpler llvm can possibly recover (i.e. effectively undo the transformation). Then if the code that uses the vector type gets more complicated, possibly because there are multiple vectors embedded in the same object, llvm might not be able to fix it anymore.

**Comment by dbeach24** [comment]
@yuyichao Thank you.  Your last comment, in particular, addresses the essence of what I'm trying to understand.
